### PR TITLE
[complex32] sum, prod : cuda only (disable jiterator reduction on windows)

### PR DESCRIPTION
**Pull request by kshitij12345**
Ref: https://github.com/pytorch/pytorch/issues/74537

### CODE REVIEW

File: aten/src/ATen/native/cuda/ReduceSumProdKernel.cu [hunk 1]
```diff
@@ -19,6 +19,33 @@ struct sum_functor {
...
+#if AT_USE_JITERATOR()
+  void operator()(TensorIterator& iter) {
+    using scalar_t = c10::complex<at::Half>;
+    std::string func = jiterator_stringify(
+    arg_t combine(arg_t a, arg_t b) {
+      return a + b;
+    }
+    );
+    jitted_gpu_reduce_kernel<sum_name, scalar_t, scalar_t>(
+        iter, func, 0.);
```
Comment (kshitij12345) [inline aten/src/ATen/native/cuda/ReduceSumProdKernel.cu:35]: Implicit upcasting occurs in jiterated code. (also `jitted_gpu_reduce_kernel` doesn't take `accumulation type`.)

File: aten/src/ATen/native/cuda/ReduceSumProdKernel.cu [hunk 2]
```diff
@@ -19,6 +19,33 @@ struct sum_functor {
...
+  }
+#else
+  void operator()(TensorIterator& iter) {
+    using scalar_t = c10::complex<at::Half>;
+    using acc_t = c10::complex<float>;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ReduceSumProdKernel.cu:40]: This should be `at::opmath_type<scalar_t>`?

Comment (kshitij12345) [reply aten/src/ATen/native/cuda/ReduceSumProdKernel.cu:40]: Yeah. Will update.

**Comment by kshitij12345** [comment]
Failures look irrelevant. Will rebase post review. PTAL @anjali411 @ngimel :) Thanks!

**Review by anjali411 — APPROVED** [review]
LGTM conditional on the CI being green

File: aten/src/ATen/native/cuda/ReduceSumProdKernel.cu [hunk 3]
```diff
@@ -19,6 +20,35 @@ struct sum_functor {
   }
 };
 
+// jiterated specialization for `complex<Half>`
+const char sum_name[] = "sum";
+template <>
+struct sum_functor<c10::complex<at::Half>> {
+// jiterator reduction fails on windows
+// Ref: https://github.com/pytorch/pytorch/issues/77305
+#if AT_USE_JITERATOR() && !defined(_MSC_VER)
```
Comment (kshitij12345) [inline aten/src/ATen/native/cuda/ReduceSumProdKernel.cu:29]: @ngimel have disabled jiterator on windows

**Pull request closed without merging**